File: services/api/models/list.js
'use strict';
module.exports = (sequelize, DataTypes) => {

  const List = sequelize.define('List', {
    title: {
      type: DataTypes.STRING,
      allowNull: false,
      validate: {
        notEmpty: true
      }
    },
    description: DataTypes.STRING,
    averageRating: DataTypes.FLOAT,
    minRating: DataTypes.INTEGER,
    maxRating: DataTypes.INTEGER
  }, {});

  List.associate = function(models) {
    List.belongsToMany(models.Movie, {through: 'ListEntry'});
    List.belongsTo(models.User);
  };

  List.statsFor = function(movies) {
    const ratings = movies.map((movie) => movie.rating).filter(x => x === 0 || x);
    if (ratings.length > 0) {
      return {
        averageRating: ratings.reduce((a, b) => a + b, 0) / ratings.length,
        minRating: Math.min(...ratings),
        maxRating: Math.max(...ratings)
      };
    } else {
      return {
        averageRating: null,
        minRating: null,
        maxRating: null
      };
    }
  }

  List.prototype.updateDetails = async function() {
    const movies = await this.getMovies();
    
    const stats = List.statsFor(movies);
    this.averageRating = stats.averageRating;
    this.minRating = stats.minRating;
    this.maxRating = stats.maxRating;
  
    if (movies.length == 0) {
      this.description = null;
    } else if (movies.length == 1) {
      this.description = movies[0].title;
    } else if (movies.length == 2) {
      this.description = `${movies[0].title} and ${movies[1].title}`;
    } else {
      const remainder = movies.length - 2;
      const remainderDescription = remainder > 1 ? 'others' : 'other';
      this.description = `${movies[0].title}, ${movies[1].title} and ${remainder} ${remainderDescription}`;
    }
  
    await this.save();
  };

  return List;
};

File: services/api-rails/app/models/list.rb
class List < ApplicationRecord
  belongs_to :user
  has_and_belongs_to_many :movies
  has_many :genres, through: :movies

  before_save :update_details

  def self.stats_for(movies)
    ratings = movies.pluck(:rating).compact
    if ratings.length > 0
      return {
          average_rating: ratings.sum.to_f / ratings.length,
          min_rating: ratings.min,
          max_rating: ratings.max
      }
    else
      return {
          average_rating: nil,
          min_rating: nil,
          max_rating: nil
      }
    end
  end

  def update_details
    stats = List.stats_for(self.movies)
    self.average_rating = stats[:average_rating]
    self.min_rating = stats[:min_rating]
    self.max_rating = stats[:max_rating]

    if self.movies.size == 0
      self.description = nil
    elsif self.movies.size == 1
      self.description = self.movies[0].title
    elsif self.movies.length == 2
      self.description = "#{self.movies[0].title} and #{self.movies[1].title}"
    else
      remainder = movies.length - 2
      remainder_description = remainder > 1 ? 'others' : 'other'
      self.description = "#{movies[0].title}, #{movies[1].title} and #{remainder} #{remainder_description}"
    end
  end
end
